marlowe-playground-client/src/Halogen/ActusBlockly.purs

```diff
@@ -0,0 +1,172 @@
+module Halogen.ActusBlockly where
+
+import Data.Json.JsonEither
+
+import Blockly (BlockDefinition, ElementId(..), getBlockById)
+import Blockly as Blockly
+import Blockly.Generator (Generator, newBlock, blockToCode)
+import Blockly.Types as BT
+import Control.Monad.Except (ExceptT(..), except, runExceptT)
+import Control.Monad.ST as ST
+import Control.Monad.ST.Ref as STRef
+import Control.Monad.State (modify_)
+import Data.Bifunctor (lmap)
+import Data.Either (Either(..), note)
+import Data.Json.JsonEither (JsonEither(..))
+import Data.Lens (Lens', _1, assign, use)
+import Data.Lens.Record (prop)
+import Data.Maybe (Maybe(..))
+import Data.Symbol (SProxy(..))
+import Effect.Class (class MonadEffect)
+import Halogen (ClassName(..), Component, HalogenM, RefLabel(..), liftEffect, mkComponent, raise)
+import Halogen as H
+import Halogen.HTML (HTML, button, div, text)
+import Halogen.HTML.Events (onClick)
+import Halogen.HTML.Properties (class_, classes, id_, ref)
+import Marlowe.ActusBlockly (buildGenerator)
+import Marlowe.Holes (Term(..))
+import Marlowe.Parser as Parser
+import Marlowe.Semantics (Contract(..))
+import Prelude (Unit, bind, const, discard, map, pure, show, unit, zero, ($), (<<<), (<>))
+import Text.Extra as Text
+import Text.Pretty (pretty)
+import Type.Proxy (Proxy(..))
+
+type BlocklyState
+  = { actusBlocklyState :: Maybe BT.BlocklyState
+    , generator :: Maybe Generator
+    , errorMessage :: Maybe String
+    }
+
+_actusBlocklyState :: Lens' BlocklyState (Maybe BT.BlocklyState)
+_actusBlocklyState = prop (SProxy :: SProxy "actusBlocklyState")
+
+_generator :: Lens' BlocklyState (Maybe Generator)
+_generator = prop (SProxy :: SProxy "generator")
+
+_errorMessage :: Lens' BlocklyState (Maybe String)
+_errorMessage = prop (SProxy :: SProxy "errorMessage")
+
+data BlocklyQuery a
+  = Resize a
+  | SetError String a
+
+data ContractFlavour = FS | F
+
+data BlocklyAction
+  = Inject String (Array BlockDefinition)
+  | GetTerms ContractFlavour
+
+data BlocklyMessage
+  = Initialized
+  | CurrentTerms ContractFlavour String
+
+type Slots
+  = ()
+
+type DSL m a
+  = HalogenM BlocklyState BlocklyAction Slots BlocklyMessage m a
+
+blockly :: forall m. MonadEffect m => String -> Array BlockDefinition -> Component HTML BlocklyQuery Unit BlocklyMessage m
+blockly rootBlockName blockDefinitions =
+  mkComponent
+    { initialState: const { actusBlocklyState: Nothing, generator: Nothing, errorMessage: Nothing }
+    , render
+    , eval:
+      H.mkEval
+        { handleQuery
+        , handleAction
+        , initialize: Just $ Inject rootBlockName blockDefinitions
+        , finalize: Nothing
+        , receive: const Nothing
+        }
+    }
+
+handleQuery :: forall m a. MonadEffect m => BlocklyQuery a -> DSL m (Maybe a)
+handleQuery (Resize next) = do
+  mState <- use _actusBlocklyState
+  case mState of
+    Just state ->
+      pure
+        $ ST.run
+            ( do
```

Comment:
we tend to not use the `()` in purescript, since `do` has an implicit separation

Reply:
That’s copy paste as well :)